### PR TITLE
Translate spoke category title directly at runtime

### PR DESCRIPTION
**Pull request by M4rtinK**
At the moment we mark spoke category title stored in class attribute for
translation and then translate it when hub sets up the category widget.

This had to be done as the class attribute is evaluated on code parsing
time and Anaconda might have a different locale (due to user selecting
installation language) by the tame the string is being displayed &
We would then translate the string at runtime.

This is unfortunately problematic for Initial Setup and addons as there
the category title will be marked for translation in the IS/addon
context and then the translation will be attempted in the Anaconda
context, which of course does not have the translated string.

So lets drop this split and instead translate the category title always
at runtime by providing it via a static class method. IS/addons can then
do the same, making sure the category title will be always translated at
the appropriate time & in their translation context.

Also clean up and document the category widget handling code a bit.

### CODE REVIEW

**Comment by M4rtinK** [comment]
Dependent fix in Initial Setup: https://github.com/rhinstaller/initial-setup/pull/110

**Review by VladimirSlavik — APPROVED** [review]
Looks good to me. Thank you!

**Review by poncovka — COMMENTED** [review]
I have suggested some small improvements. Otherwise, it looks great. Thanks!

**Comment by poncovka** [comment]
Could you open a pull request for the hello world addon, please?

**Comment by VladimirSlavik** [comment]
You will need to rebase for tests, I think...

File: pyanaconda/ui/categories/customization.py
Comment (M4rtinK) [reply pyanaconda/ui/categories/customization.py:28]: Yep, lets change this.

Hmm, I though I saw we were doing this somewhere else and did the same but 
can't find it now.

File: pyanaconda/ui/categories/__init__.py
Comment (M4rtinK) [reply pyanaconda/ui/categories/__init__.py:58]: Sure.

BTW, should it be always done ? I see some places we just have pass for an abstract method:

https://github.com/rhinstaller/anaconda/blob/master/pyanaconda/modules/payloads/payload/payload_base.py#L47

Or is it somehow related to these being properties ?

File: pyanaconda/ui/categories/__init__.py
Comment (M4rtinK) [reply pyanaconda/ui/categories/__init__.py:46]: Changed.

**Comment by M4rtinK** [comment]
> Could you open a pull request for the hello world addon, please?

Will do. :)

File: pyanaconda/ui/categories/__init__.py
Comment (M4rtinK) [reply pyanaconda/ui/categories/__init__.py:41]: Good point, changed.

File: pyanaconda/ui/categories/__init__.py
Comment (poncovka) [reply pyanaconda/ui/categories/__init__.py]: From the functionality point of view, it doesn't matter. However, I think that the code is more readable that way.

**Review by poncovka — APPROVED** [review]
Looks good to me. Thanks!

**Comment by M4rtinK** [comment]
PR adjusting category handling  in Hello World Addon: https://github.com/rhinstaller/hello-world-anaconda-addon/pull/22

**Review by jkonecny12 — COMMENTED** [review]
As I wrote elsewhere. This seems fine from the code PoV but are we sure about switching ``N_`` to ``_``? This won't work when compare is used. 

Also thinking that we may change the ``get_title()`` to ``get_localized_title()`` to avoid issues with using this in the future. The problem here is that it will work for English but for nothing else.

**Comment by poncovka** [comment]
> As I wrote elsewhere. This seems fine from the code PoV but are we sure about switching `N_` to `_`? This won't work when compare is used.

The goal of this change is to use `_`, because addons are able to translate their own strings with the right context. Anaconda translates  `N_` strings with its own context and it results in untranslated strings.

I don't understand why you are worried about comparison of titles of categories. Categories are sorted by their sort order. This issue will be valid for spokes, but that's for another pull request.

> Also thinking that we may change the `get_title()` to `get_localized_title()` to avoid issues with using this in the future. The problem here is that it will work for English but for nothing else.

What kind of issues?

**Comment by poncovka** [comment]
My bad. We sort categories by titles in TUI. However, I think we should use the sort order as well.

**Comment by jkonecny12** [comment]
> > As I wrote elsewhere. This seems fine from the code PoV but are we sure about switching `N_` to `_`? This won't work when compare is used.
> 
> The goal of this change is to use `_`, because addons are able to translate their own strings with the right context. Anaconda translates `N_` strings with its own context and it results in untranslated strings.
> 
> I don't understand why you are worried about comparison of titles of categories. Categories are sorted by their sort order. This issue will be valid for spokes, but that's for another pull request.
> 
> > Also thinking that we may change the `get_title()` to `get_localized_title()` to avoid issues with using this in the future. The problem here is that it will work for English but for nothing else.
> 
> What kind of issues?

I mean that it's pretty easy to use this for something like:
```
if spoke.get_title() == "SourceSpoke":
```
without realizing that it won't work on non-english locale.

**Comment by poncovka** [comment]
> I mean that it's pretty easy to use this for something like:
> 
> ```
> if spoke.get_title() == "SourceSpoke":
> ```
> 
> without realizing that it won't work on non-english locale.

The title will be `Installation Source` in this case. I think it is pretty easy to spot this kind of error.

**Comment by VladimirSlavik** [comment]
It makes sense that translation should be mentioned in the method name. Otherwise I think this might be fine. Sorting by name which is untranslated-but-should-be sounds mostly not ok, so if this makes a fix for that necessary, I'd still consider that a good thing.

**Comment by M4rtinK** [comment]
So to summarize the current issues & possible solutions (**for both spokes and categories to have the discussion in one place**):

* `get_title()` vs `get_localized_title()`
Personally, having the `get_localized_title()` does not make sense to me unless we also provide `get_title()` that returns the untranslated string. If we can avoid that, documenting get_title() returns translated strings should be IMHO enough.

* comparisons on category/spoke titles
These should not be a thing at all - as any other attempts to match on human readable strings in the GUI. If the human readable string is translated or not IMHO does not make a difference here.

For robust comparisons some machine readable spoke descriptor should be used - possibly class name or something new we can introduce similar to the `help_id` we already have.

* translations vs category & spoke ordering
I'm a bit confused at the moment if we are all taking about the same problem. 

If I understand things correctly, the current status quo is that we sort the spokes and (categories in TUI) based on their untranslated English name. This has the benefit of spokes (and categories in TUI) being always in the same place in the UI regardless of the selected locale. On the other hand, users that are not native in English might be confused why the spokes are not ordered according to their expected native alphabet ordering.

With change in spoke (and category localization) we might be be able to sort the spokes (and categories) in TUI based on the correct native alphabet ordering (as long as we make sure to only do it once locale has been set in the welcome spoke).  The downside here is that spokes might be located in different parts of the hub based on currently selected language. This can make debugging of l10n related issues as one can't simply put things side by side as before, but navigate based on spoke icon, which will be the same. I can imagine this also might momentarily break GUI testing tools that expect a certain spoke order and/or work based on screenshot bitmaps.

* inconsistency in category ordering
GUI orders categories by `sortOrder` but TUI seems to ignore it ? Do we want (or need ?) to keep it this way ? Or should we make the TUI `sortOder` aware to make this consistent ?

If there is a good reason we might want to sort the categories in a different way in the GUI and TUI I guess we could have a `sortOrderTUI` property if really needed.

* do we actually want to sort spokes alphabetically ?

As mentioned above it has some serious pitfalls & it might make sense to manually set the sort order to both make it consistent as well as possible order the spokes on the hub in a more ergonomic way. I remember @andreasn recommended an improved spoke & category ordering layout but I'm afraid it might have got lost in all the install-time-subscription related work & I can't find it right now.

**Comment by VladimirSlavik** [comment]
Thanks for the summary, I agree with most of what's there.

The order should reflect something logical, eg. how people will move through these, not random, such as first letter ;-) At least that's what seems sane to me. Following from that, I completely agree that comparisons should not be based on strings.

**Comment by jkonecny12** [comment]
> > I mean that it's pretty easy to use this for something like:
> > ```
> > if spoke.get_title() == "SourceSpoke":
> > ```
> > 
> > 
> > without realizing that it won't work on non-english locale.
> 
> The title will be `Installation Source` in this case. I think it is pretty easy to spot this kind of error.

OK bad example, the point is that for "CS" localization it will not be 'Installation Source' anymore.

**Comment by jkonecny12** [comment]
> So to summarize the current issues & possible solutions (**for both spokes and categories to have the discussion in one place**):
> ### `get_title()` vs `get_localized_title()`
> 
> Personally, having the `get_localized_title()` does not make sense to me unless we also provide `get_title()` that returns the untranslated string. If we can avoid that, documenting get_title() returns translated strings should be IMHO enough.

I don't think there is a reason to have both. If we don't require to have ``get_title()`` then I don't see a reason to have it. On the other hand I still think that the name should explicitly tell that the string is already localized.

> ### comparisons on category/spoke titles
> 
> These should not be a thing at all - as any other attempts to match on human readable strings in the GUI. If the human readable string is translated or not IMHO does not make a difference here.
> 
> For robust comparisons some machine readable spoke descriptor should be used - possibly class name or something new we can introduce similar to the `help_id` we already have.

Good point on that but I still think the same about the above. Even though, I won't block this PR because of it.

**Comment by M4rtinK** [comment]
After a meeting with all interested parties, we have reached the following conclusion in regards to the spoke related localization and attribute fixes/cleanups:

* `sort_order()` abstract property will be added to spokes, to provide deterministic ordering of spokes
* initial spoke sort order will be based on current alphabetical ordering (English) spoke titles, in the future the sort order can be adjusted as necessary
* category sorting in TUI will also be based on sort order instead of current alphabetical sort (GUI already displays categories based on sort order)
* `get_title()` will be used for category titles and `title()` for spokes, with the documentation clearly mentioning the output is expected to be localized (as is output from `status()` already)

**Comment by andreasn** [comment]
@M4rtinK maybe it was this mockup?
Been a while since I last looked at this, but what I was attempting to solve at the time was that some spokes were dependant on others. So _Connect to Red Hat_ was dependant on the network being turned on (off by default in RHEL) in _Network & Host name_ and _Software selection_ dependant on _Installation source_ being set up etc.

image

Didn't look to deep into the problem though, so not sure how crucial it is.

**Comment by M4rtinK** [comment]
@andreasn Thanks, that's the mockup I meant (and failed to find myself :P)! And it's definitely not a pressing concern at the moment - it's just that once this PR & #2960 are merged, we will actually have full controls over category and spoke placement in both GUI & TUI. Then we can actually do something about it - rest assured we will loop you in once that time comes. :)

**Comment by M4rtinK** [comment]
@poncovka @jkonecny12 @VladimirSlavik So I did some final testing with TUI category sort based on sort order.

This is how it looked before:
TUI_pre_localization_fixes

This is how it looks now:
TUI_post_localization_fixes_category_sort

GUI also looks good:
GUI_post_localization_fixes_category_sort

So I think we can merge this (and the IS & Hello World Addon PRs), so that #2960 can be rebase on top of it & there is less change at one point in time.

File: pyanaconda/ui/common.py
Comment (poncovka) [inline pyanaconda/ui/common.py:704]: It should be sorted by `__class__.__name__`.

File: pyanaconda/ui/common.py
Comment (poncovka) [inline pyanaconda/ui/common.py:668]: This sort doesn't seem to be useful.

File: tests/nosetests/pyanaconda_tests/common_ui_code_tests.py
Comment (poncovka) [inline tests/nosetests/pyanaconda_tests/common_ui_code_tests.py:66]: A and B is already in the input.

**Review by poncovka — COMMENTED** [review]
I would prefer if this pull request wasn't merged without a final review.

File: pyanaconda/ui/common.py
Comment (poncovka) [reply pyanaconda/ui/common.py:704]: I couldn't figure out why your test is not failing, because classes are not comparable. The problem is that categories are not instances, but classes. Therefore `i` is a class and `i.__class__` is `type`. And `type` is comparable. Anyway, categories should be sorted by `__name__`. Please, make sure that your tests are testing this use case.

**Comment by M4rtinK** [comment]
@poncovka Sorry for that, I've mentioned my intent to merge the PR on our daily standup & the changes in the send seemed quite minor. I guess pinging the reviewers on the PR would work better in the end though.

File: pyanaconda/ui/common.py
Comment (M4rtinK) [reply pyanaconda/ui/common.py:668]: I assumed there is some reason we sort the categories when they are being collected (stable listing in logs ?) so I just kept the logic the same. If you think it's safe to drop this, I can definitely do that. :)

File: pyanaconda/ui/common.py
Comment (M4rtinK) [reply pyanaconda/ui/common.py:704]: Oops! Yeah, that's really wrong. :P I'll fix it up in a folowup PR.

File: tests/nosetests/pyanaconda_tests/common_ui_code_tests.py
Comment (M4rtinK) [reply tests/nosetests/pyanaconda_tests/common_ui_code_tests.py:66]: Good point, I'll change the sort order or else, it will indeed just be kept the same in output, making the test kinda pohtless.

File: pyanaconda/ui/common.py
Comment (poncovka) [reply pyanaconda/ui/common.py:668]: I would try to remove drop that. It is just confusing.

**Pull request merged**
